httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/gss/GssSchemeBase.java

```diff
@@ -0,0 +1,394 @@
+                        }
+                        state = State.SUCCEEDED;
+                        break;
+                    } else {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext = HttpClientContext.cast(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} Did not receive required challenge and GssConfig.ignoreMissingToken is false.",
+                                exchangeId);
+                        }
+                        state = State.FAILED;
+                        throw new AuthenticationException(
+                                "Did not receive required challenge and GssConfig.ignoreMissingToken is false.");
+                    }
+                }
+                queuedToken = generateToken(challengeToken, KERBEROS_SCHEME, gssHostname);
+                if (challenged) {
+                    state = State.TOKEN_READY;
+                } else if (!gssContext.isEstablished()) {
+                    if (LOG.isDebugEnabled()) {
+                        final HttpClientContext clientContext = HttpClientContext.cast(context);
+                        final String exchangeId = clientContext.getExchangeId();
+                        LOG.debug("{} GSSContext is not established.", exchangeId);
+                    }
+                    state = State.FAILED;
+                    // TODO should we have specific exception(s) for these ?
+                    throw new AuthenticationException(
+                            "GSSContext is not established.");
+                } else if (!gssContext.getMutualAuthState()) {
+                    if (requireMutualAuth) {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext = HttpClientContext.cast(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} requireMutualAuth is true but GSSContext mutualAuthState is false",
+                                exchangeId);
+                        }
+                        state = State.FAILED;
+                        throw new AuthenticationException(
+                                "requireMutualAuth is true but GSSContext mutualAuthState is false");
+                    } else {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext = HttpClientContext.cast(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} GSSContext MutualAuthState is false, but continuing because GssConfig.requireMutualAuth is false.",
+                                exchangeId);
+                        }
+                        state = State.FAILED;
```

Comment:
I do not fully understand this. If mutual auth isn't required, why fail if it is not available?

Reply:
This is a bug, I'm fixing it. (We don't have a test case for this path)

Reply:
OK